### PR TITLE
Added latest version check added

### PR DESCRIPTION
**Pull request by evalsocket**
# TL;DR
- Added log for new flytectl version
- Ignore error in control plan API call (Currently version command failed if control plan is not available). For getting the error user need to unable debugging 
- Added more unit test 

```bash
$ ➜ bin/flytectl version
A newer version of flytectl is available [v0.1.20] Please upgrade using - https://docs.flyte.org/projects/flytectl/en/latest/index.html
{
  "App": "flytectl",
  "Build": "f56ec98",
  "Version": "v0.1.12-2-gf56ec98",
  "BuildTime": "2021-06-20 17:59:23.835983 +0530 IST m=+0.014146333"
}{
  "App": "controlPlane",
  "Build": "821d232",
  "Version": "v0.6.0",
  "BuildTime": "2021-06-20 08:02:40.3949435 +0000 UTC m=+0.041891601"
}

$ ➜ bin/flytectl version (If an admin is not available)
A newer version of flytectl is available [v0.1.20] Please upgrade using - https://docs.flyte.org/projects/flytectl/en/latest/index.html
{
  "App": "flytectl",
  "Build": "f56ec98",
  "Version": "v0.1.12-2-gf56ec98",
  "BuildTime": "2021-06-20 18:00:28.863216 +0530 IST m=+0.015169421"
}
```

## Type
 - [x] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/1168

## Follow-up issue
_NA_
OR
_https://github.com/lyft/flyte/issues/<number>_

### CODE REVIEW

File: cmd/version/version.go [hunk 1]
```diff
@@ -60,23 +66,62 @@ func getVersion(ctx context.Context, args []string, cmdCtx cmdCore.CommandContex
...
+
+func compareVersion(latest, current string) string {
+	latestVersion, _ := hversion.NewVersion(latest)
+	fversion, _ := hversion.NewVersion(current)
+
+	if fversion.LessThan(latestVersion) {
+		return fmt.Sprintf("A newer version of flytectl is available [%v] Please upgrade using - https://docs.flyte.org/projects/flytectl/en/latest/index.html", latest)
+	}
+
+	return "Installed flytectl version is the latest"
```
Comment (pmahindrakar-oss) [inline cmd/version/version.go:90]: const

File: cmd/version/version.go [hunk 2]
```diff
@@ -60,23 +66,62 @@ func getVersion(ctx context.Context, args []string, cmdCtx cmdCore.CommandContex
...
+	if err != nil || v == nil {
+		logger.Debugf(ctx, "Failed to get version of control plane %v: \n", err)
+		return
+	}
 	// Print Flyteadmin
 	if err := printVersion(versionOutput{
 		Build:     v.ControlPlaneVersion.Build,
 		BuildTime: v.ControlPlaneVersion.BuildTime,
 		Version:   v.ControlPlaneVersion.Version,
 		App:       "controlPlane",
```
Comment (pmahindrakar-oss) [inline cmd/version/version.go:104]: Same here app constant

File: cmd/version/version.go [hunk 3]
```diff
@@ -60,23 +66,62 @@ func getVersion(ctx context.Context, args []string, cmdCtx cmdCore.CommandContex
...
+	data, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return "", err
+	}
+	var result = make(map[string]interface{})
+	err = json.Unmarshal(data, &result)
+	if err != nil {
+		return "", err
+	}
+	return result["tag_name"].(string), nil
```
Comment (pmahindrakar-oss) [inline cmd/version/version.go:126]: key check before converting it to string to avoid crash

File: cmd/version/version.go [hunk 4]
```diff
@@ -60,23 +66,62 @@ func getVersion(ctx context.Context, args []string, cmdCtx cmdCore.CommandContex
...
 	}
-	fmt.Print(string(b))
-	return nil
+	defer response.Body.Close()
+
+	data, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return "", err
+	}
+	var result = make(map[string]interface{})
```
Comment (pmahindrakar-oss) [inline cmd/version/version.go:121]: Can this not be map[string]string. Do we expect any other type for any of the Version responses?

File: cmd/version/version.go [hunk 5]
```diff
@@ -60,23 +66,62 @@ func getVersion(ctx context.Context, args []string, cmdCtx cmdCore.CommandContex
...
+	fmt.Print(string(b))
+	return nil
+}
+
+func compareVersion(latest, current string) string {
+	latestVersion, _ := hversion.NewVersion(latest)
+	fversion, _ := hversion.NewVersion(current)
+
+	if fversion.LessThan(latestVersion) {
+		return fmt.Sprintf("A newer version of flytectl is available [%v] Please upgrade using - https://docs.flyte.org/projects/flytectl/en/latest/index.html", latest)
```
Comment (pmahindrakar-oss) [inline cmd/version/version.go:87]: We should think of having a single step upgrade for the user by just giving him a prompt if he wants to upgrade 

check this out  https://equinox.io/ and see if its possible to use it.

File: cmd/version/version.go [hunk 6]
```diff
@@ -60,23 +66,62 @@ func getVersion(ctx context.Context, args []string, cmdCtx cmdCore.CommandContex
...
 	}
 
+	getControlePlaneVersion(ctx, cmdCtx)
+	return nil
+}
+
+func printVersion(response versionOutput) error {
+	b, err := json.MarshalIndent(response, "", "  ")
+	if err != nil {
+		return fmt.Errorf("err %v: ", err)
```
Comment (pmahindrakar-oss) [inline cmd/version/version.go:76]: Unit test covering this branch and other branches in this file

File: cmd/version/version.go [hunk 7]
```diff
@@ -22,6 +26,8 @@ Example version.
 `
 )
 
+var githubBaseURL = "https://api.github.com"
```
[the remark this answers is not in the document]
Comment (evalsocket) [reply cmd/version/version.go:29]: I used it for testing multiple test cases. I will change its value in the test. I am not using direct URL as a variable bc it gives me a  lint error

File: cmd/version/version.go (hunk 5 above)
Comment (evalsocket) [reply cmd/version/version.go:87]: Yes, I know about it and it's amazing if we can use it. You can create an issue for that

File: cmd/version/version.go (hunk 4 above)
Comment (evalsocket) [reply cmd/version/version.go:121]: ok i will create proper struct

File: cmd/version/version.go (hunk 6 above)
Comment (evalsocket) [reply cmd/version/version.go:76]: meaning ?

File: cmd/version/version.go [hunk 8]
```diff
@@ -44,39 +52,70 @@ func GetVersionCommand(rootCmd *cobra.Command) map[string]cmdCore.CommandEntry {
...
+	b, err := json.MarshalIndent(response, "", "  ")
+	if err != nil {
+		return err
+	}
+	fmt.Print(string(b))
+	return nil
+}
+
+func compareVersion(latest, current string) string {
+	semanticVersion, _ := hversion.NewVersion(latest)
```
Comment (kumare3) [inline cmd/version/version.go:87]: Handle error, maybe ignore, but check and log?

**Review by kumare3 — DISMISSED** [review]
Awesome

**Review by pmahindrakar-oss — APPROVED** [review]
